nano/node/node.cpp

```diff
@@ -2937,80 +2956,75 @@ bool nano::block_arrival::recent (nano::block_hash const & hash_a)
-	std::lock_guard<std::mutex> lock (mutex);
-	auto now (std::chrono::steady_clock::now ());
-	auto transaction (node.store.tx_begin_read ());
-	auto current (reps.begin ());
-	while (current != reps.end () && current->last_heard + std::chrono::seconds (nano::node::cutoff) < now)
+	auto transaction (ledger.store.tx_begin_read ());
+	if (ledger.weight (transaction, rep_a) > nano::Gxrb_ratio)
 	{
-		auto old_stake (online_stake_total);
-		online_stake_total -= node.ledger.weight (transaction, current->representative);
-		if (online_stake_total > old_stake)
-		{
-			// underflow
-			online_stake_total = 0;
-		}
-		current = reps.erase (current);
+		std::lock_guard<std::mutex> lock (mutex);
+		reps.insert (rep_a);
 	}
-	auto rep_it (reps.get<1> ().find (rep));
-	auto info (nano::rep_last_heard_info{ now, rep });
-	if (rep_it == reps.get<1> ().end ())
+}
+
+void nano::online_reps::sample ()
+{
+	auto transaction (ledger.store.tx_begin_write ());
+	// Discard oldest entries
+	while (ledger.store.online_weight_count (transaction) >= weight_samples)
 	{
-		auto old_stake (online_stake_total);
-		online_stake_total += node.ledger.weight (transaction, rep);
-		if (online_stake_total < old_stake)
-		{
-			// overflow
-			online_stake_total = std::numeric_limits<nano::uint128_t>::max ();
-		}
-		reps.insert (info);
+		auto oldest (ledger.store.online_weight_begin (transaction));
+		assert (oldest != ledger.store.online_weight_end ());
+		ledger.store.online_weight_del (transaction, oldest->first);
 	}
-	else
+	// Calculate current active rep weight
+	nano::uint128_t current;
+	for (auto & i : reps)
 	{
-		reps.get<1> ().replace (rep_it, info);
+		current += ledger.weight (transaction, i);
 	}
+	reps.clear ();
+	ledger.store.online_weight_put (transaction, std::chrono::system_clock::now ().time_since_epoch ().count (), current);
+	auto trend_l (trend (transaction));
+	std::lock_guard<std::mutex> lock (mutex);
+	online = trend_l;
 }
 
-void nano::online_reps::recalculate_stake ()
+nano::uint128_t nano::online_reps::trend (nano::transaction & transaction_a)
 {
-	std::lock_guard<std::mutex> lock (mutex);
-	online_stake_total = 0;
-	auto transaction (node.store.tx_begin_read ());
-	for (auto it : reps)
+	std::vector<nano::uint128_t> items;
+	items.reserve (weight_samples + 1);
+	items.push_back (minimum);
+	for (auto i (ledger.store.online_weight_begin (transaction_a)), n (ledger.store.online_weight_end ()); i != n; ++i)
 	{
-		online_stake_total += node.ledger.weight (transaction, it.representative);
+		items.push_back (i->second.number ());
 	}
-	auto now (std::chrono::steady_clock::now ());
-	std::weak_ptr<nano::node> node_w (node.shared ());
-	node.alarm.add (now + std::chrono::minutes (5), [node_w]() {
-		if (auto node_l = node_w.lock ())
-		{
-			node_l->online_reps.recalculate_stake ();
-		}
-	});
+	std::sort (items.begin (), items.end ());
+	// Pick median value for our target vote weight
+	nano::uint128_t result (items[items.size () / 2]);
```

Comment:
`std:nth_element` can be used which finds what a certain index's value would be if the whole collection is sorted (but is much quicker):
```
    auto i = items.size() / 2;
    nth_element (items.begin (), items.begin () + i, items.end ());
    return nano::uint128_t{ items[i] };
```

Reply:
Done